platform/openide.filesystems/src/org/openide/filesystems/FileUtil.java

```diff
@@ -2323,17 +2324,24 @@ private static boolean isArchiveFileImpl(final URL url, final boolean strict) {
     }
 
     private static Iterable<? extends ArchiveRootProvider> getArchiveRootProviders() {
-        Lookup.Result<ArchiveRootProvider> res = archiveRootProviders.get();
-        if (res == null) {
-            res = new ProxyLookup(
-                Lookups.singleton(new JarArchiveRootProvider()),
-                Lookup.getDefault()).lookupResult(ArchiveRootProvider.class);
-            if (!archiveRootProviders.compareAndSet(null, res)) {
-                res = archiveRootProviders.get();
+        if (archiveRootProviderCache == null) {
+            Lookup.Result<ArchiveRootProvider> res = archiveRootProviders.get();
+            if (res == null) {
+                res = new ProxyLookup(
+                    Lookups.singleton(new JarArchiveRootProvider()),
+                    Lookup.getDefault()).lookupResult(ArchiveRootProvider.class);
+                if (!archiveRootProviders.compareAndSet(null, res)) {
+                    res = archiveRootProviders.get();
+                    res.addLookupListener((ev) -> {
+                        archiveRootProviderCache = null;
+                    });
+                }
             }
+            archiveRootProviderCache = new LinkedList<>(res.allInstances());
         }
-        return res.allInstances();
+        return archiveRootProviderCache;
```

Comment:
Why is this safe? If I'm reading it correctly after the first query no new ArchiveRootProvider will be found. So If at an unfortunate moment this is called before all ArchiveRootProviders are found (or they are loaded at runtime) this will return invalid results.

Reply:
In Line 2335 we add a listener to the lookup result, so if new Archive providers would be registered in the system the cache would be invalidated and it's content would be re-read. Also so far there are 2 ArchiveRootProvider implementations in the system one for jars and one for jrt-s.
Profile snapshots attached to NETBEANS-2940 shows that NetBeans can spend considerable amount of time searching the non-jar provider almost every time when an archive root has to be evaluated.

Reply:
Thank you for the explanation - indeed I overlooked the lookup listener and its invalidation implementation. I toyed a bit with the ProxyLookup and it looks to react correctly.
Looks sane to me now.

Reply:
Thank you for the review! Let's see how does it work in the real life, then.